avm/src/vm/opcode.rs

```diff
@@ -3831,6 +3831,339 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
+            HandlerMemory::transfer(&ds, 0, &mut hm, CLOSURE_ARG_MEM_START + 1)?;
+            hm.register_out(args[0], 1, CLOSURE_ARG_MEM_START + 2)?;
+            let subhandler = HandlerFragment::new(args[1], 0);
+            let hm = subhandler.run(hm).await?;
+            let hm = hm.drop_parent()?;
+            HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START, &mut hand_mem, CLOSURE_ARG_MEM_START)?;
+            hand_mem.push_fixed(args[2], 1i64)?;
+            hand_mem.push_register(args[2], CLOSURE_ARG_MEM_START)?;
+            // Also grab the mutation to the datastore value and re-insert it
+            let mut newds = HandlerMemory::new(None, 1)?;
+            HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START + 1, &mut newds, 0)?;
+            drop(ds);
+            DS.insert(nskey, newds);
+          },
+          None => {
+            hand_mem.push_fixed(args[2], 0i64)?;
+            hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal("namespace-key pair not found"))?;
+          },
+        }
+      } else {
+        // TODO
+      }
+      Ok(hand_mem)
+    })
+  });
+  unpred_cpu!(dsmonly => fn(args, hand_mem) {
+    Box::pin(async move {
+      let nsref = hand_mem.read_fractal(args[0])?;
+      let ns = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 0).0)?;
+      let key = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 1).0)?;
+      let nskey = format!("{}:{}", ns, key);
+      let ctrl_port = CONTROL_PORT_CHANNEL.get();
+      let ctrl_port = match ctrl_port {
+        Some(ctrl_port) => Some(ctrl_port.borrow().clone()), // TODO: Use thread-local storage
+        None => None,
+      };
+      let is_key_owner = match ctrl_port {
+        Some(ref ctrl_port) => ctrl_port.is_key_owner(&nskey),
+        None => true,
+      };
+      if is_key_owner {
+        let maybe_hm = DS.get(&nskey);
+        if let Some(ds) = maybe_hm {
+          let mut hm = HandlerMemory::fork(hand_mem.clone())?;
+          HandlerMemory::transfer(&ds, 0, &mut hm, CLOSURE_ARG_MEM_START + 1)?;
+          let subhandler = HandlerFragment::new(args[1], 0);
+          let hm = subhandler.run(hm).await?;
+          let hm = hm.drop_parent()?;
+          // Grab the mutation to the datastore value and re-insert it
+          let mut newds = HandlerMemory::new(None, 1)?;
+          HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START + 1, &mut newds, 0)?;
+          drop(ds);
+          DS.insert(nskey, newds);
+        }
+      } else {
+        // TODO
+      }
+      Ok(hand_mem)
+    })
+  });
+  unpred_cpu!(dswonly => fn(args, hand_mem) {
+    Box::pin(async move {
+      let with = hand_mem.read_fractal(args[0])?;
+      let nsref = hand_mem.read_from_fractal(&with, 0).0;
+      let ns = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 0).0)?;
+      let key = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 1).0)?;
+      let nskey = format!("{}:{}", ns, key);
```

Comment:
And why the `register_out` it is not necessary here?

Reply:
I did it [here instead](https://github.com/alantech/alan/pull/596/files#diff-aa0e5a81bac19bdacb5f06f8d0cefe2516b53e02227ac03e058f2230d02f7ba2R4066)

It shouldn't affect things either way. They're very similar (since they all do nearly the same things with slight differences) but I wrote them separately and apparently put all of the `HandlerMemory` mutation down in the `if is_key_owner` block, which is probably the better way to do this?

Reply:
Yes, my confusion was that in `dsrwith` for instance you `register_out` the original `hand_mem` (before the fork) and later on you `register_out` the forked `hm`, while in this opcode you only `register_out` the forked one.